### PR TITLE
fix(node): Use upstream implementation for streams

### PR DESCRIPTION
**Pull request by Soremwar**
New approach, using implementation directly adapted from Node codebase

Note: Not ready, but can't seem to be able to mark the PR as draft :upside_down_face:

### CODE REVIEW

**Comment by Soremwar** [comment]
I really want to understand what was the intent here, but all I see is a recursive function with no shortcircuit

image

**Comment by kt3k** [comment]
@Soremwar 
Thanks for the huge work!

> but all I see is a recursive function with no shortcircuit

The node.js source code uses `FunctionPrototypeSymbolHasInstance(this, object)` expression. I think this is interpreted as `Function.prototype[Symbol.hasInstance].call(this, object)` in our code base, which isn't a recursive call in my view.

File: node/internal/streams/destroy.js [hunk 1]
```diff
@@ -0,0 +1,320 @@
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+// deno-lint-ignore-file
+
+import { aggregateTwoErrors } from "../errors.js";
+import { ERR_MULTIPLE_CALLBACK } from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/destroy.js:6]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/duplex.js [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+import {
+  isDuplexNodeStream,
+  isIterable,
+  isNodeStream,
+  isReadable,
+  isReadableNodeStream,
+  isWritable,
+  isWritableNodeStream,
+} from "./utils.js";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/duplex.js:21]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/end-of-stream.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+// deno-lint-ignore-file
+
+import { AbortError, ERR_STREAM_PREMATURE_CLOSE } from "../../_errors.ts";
+import { once } from "../util.ts";
+import {
+  validateAbortSignal,
+  validateFunction,
+  validateObject,
+} from "../validators.js";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/end-of-stream.js:11]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/from.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+// deno-lint-ignore-file
+
+import { Buffer } from "../../buffer.ts";
+import { ERR_INVALID_ARG_TYPE, ERR_STREAM_NULL_VALUES } from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/from.js:6]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/pipeline.js [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+import { once } from "../util.ts";
+import { stderr, stdout } from "./readable.js";
+import { validateAbortSignal, validateCallback } from "../validators.js";
+import {
+  ERR_INVALID_ARG_TYPE,
+  ERR_INVALID_RETURN_VALUE,
+  ERR_MISSING_ARGS,
+  ERR_STREAM_DESTROYED,
+} from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/pipeline.js:15]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/readable.js [hunk 6]
```diff
@@ -0,0 +1,1455 @@
...
+import { StringDecoder } from "../../string_decoder.ts";
+import { validateObject } from "../validators.js";
+import { Writable } from "./writable.js";
+import {
+  ERR_INVALID_ARG_TYPE,
+  ERR_METHOD_NOT_IMPLEMENTED,
+  ERR_STREAM_PUSH_AFTER_EOF,
+  ERR_STREAM_UNSHIFT_AFTER_END_EVENT,
+} from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/readable.js:17]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/transform.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+// deno-lint-ignore-file
+
+import { ERR_METHOD_NOT_IMPLEMENTED } from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/transform.js:5]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

File: node/internal/streams/writable.js [hunk 8]
```diff
@@ -0,0 +1,879 @@
...
+  ERR_METHOD_NOT_IMPLEMENTED,
+  ERR_MULTIPLE_CALLBACK,
+  ERR_STREAM_ALREADY_FINISHED,
+  ERR_STREAM_CANNOT_PIPE,
+  ERR_STREAM_DESTROYED,
+  ERR_STREAM_NULL_VALUES,
+  ERR_STREAM_WRITE_AFTER_END,
+  ERR_UNKNOWN_ENCODING,
+} from "../../_errors.ts";
+import * as process from "../../_process/process.ts";
```
Comment (bartlomieju) [inline node/internal/streams/writable.js:21]:
```suggestion
import { nextTick } from "../../_next_tick.ts";
```

**Review by bartlomieju — COMMENTED** [review]
Looking good

**Comment by bartlomieju** [comment]
@Soremwar I will open a PR with typing to your fork repo.

**Comment by Soremwar** [comment]
@bartlomieju Done that way so we have as little changes to the source as possible. In case we get a bug I want to be able to compare this to the original source and know for sure the only changes happening are in the imports and exports of the file

**Comment by bartlomieju** [comment]
> @bartlomieju Done that way so we have as little changes to the source as possible. In case we get a bug I want to be able to compare this to the original source and know for sure the only changes happening are in the imports and exports of the file

I understand, but won't that lead to the circular imports problem anyway?

**Comment by Soremwar** [comment]
Fortunately not cause:
```
// The following are all the process APIs that don't depend on the stream module
// They have to be split this way to prevent a circular dependency
```

**Comment by Soremwar** [comment]
News flash, tests are passing

However since `streams` started working as they should `net` has kinda janky, and I'm not sure if it's due to an implementation that took into account previous stream quirks or something else I unintentionally changed

@bartlomieju Please advise

File: node/_tools/config.json [hunk 9]
```diff
@@ -164,71 +129,6 @@
...
-      "test-net-server-close.js",
-      "test-net-server-listen-options-signal.js",
-      "test-net-server-listen-options.js",
-      "test-net-server-listen-remove-callback.js",
-      "test-net-server-max-connections.js",
-      "test-net-server-options.js",
-      "test-net-server-pause-on-connect.js",
-      "test-net-server-try-ports.js",
-      "test-net-socket-connect-without-cb.js",
-      "test-net-socket-connecting.js",
```
Comment (bartlomieju) [inline node/_tools/config.json:217]: Why do we need to disable these tests?

**Comment by bartlomieju** [comment]
> However since streams started working as they should net has kinda janky, and I'm not sure if it's due to an implementation that took into account previous stream quirks or something else I unintentionally changed

I presume it's the former. I see the CI is green, are you talking about Node native tests or the `*_test.ts` files?

**Comment by Soremwar** [comment]
> I presume it's the former. I see the CI is green, are you talking about Node native tests or the *_test.ts files?
Node native tests for net, those are the ones that are making the CI fail, that's why I disabled them

**Comment by bartlomieju** [comment]
> > I presume it's the former. I see the CI is green, are you talking about Node native tests or the *_test.ts files?
> 
> Node native tests for net, those are the ones that are making the CI fail, that's why I disabled them

Yikes, that's pretty bad. I think we should address these problem in this PR. I can look into that if you want. Please revert disabled tests.

Side note: I tried #1554 on top of this branch, unfortunately it still doesn't work as expected. I guess it might be an error with how we handle stdin after all.

**Comment by Soremwar** [comment]
> Side note: I tried #1554 on top of this branch, unfortunately it still doesn't work as expected. I guess it might be an error with how we handle stdin after all.

Might be the way we build stdin as a stream
https://github.com/denoland/deno_std/blob/d960e155c1f54b738708201bc338c2dca27e1651/node/_process/streams.ts#L34-L62

**Comment by Soremwar** [comment]
> Yikes, that's pretty bad. I think we should address these problem in this PR. I can look into that if you want. Please revert disabled tests.

Will do

**Comment by bartlomieju** [comment]
> > Side note: I tried #1554 on top of this branch, unfortunately it still doesn't work as expected. I guess it might be an error with how we handle stdin after all.
> 
> Might be the way we build stdin as a stream
> 
> https://github.com/denoland/deno_std/blob/d960e155c1f54b738708201bc338c2dca27e1651/node/_process/streams.ts#L34-L62

Sorry, my bad, it was a bad merge on my side. I fixed the merge and now it works as expected. Nice work @Soremwar

**Comment by Soremwar** [comment]
> Sorry, my bad, it was a bad merge on my side. I fixed the merge and now it works as expected. Nice work @Soremwar
Score!

**Comment by Soremwar** [comment]
@bartlomieju Tests for streams are clean and passing now

Out of personal preference I'd rather have the net problems handled in a different PR, since this one is already quite big

**Comment by bartlomieju** [comment]
@Soremwar that's reasonable let's do it. Though it seems the tests hanged

**Comment by Soremwar** [comment]
They hanged cause I returned the net ones

**Comment by bartlomieju** [comment]
I see. Okay let's land previous commitment and I can work on a fix tomorrow. Sorry for commotion.

**Comment by Soremwar** [comment]
Done, ready for review

**Review by bartlomieju — APPROVED** [review]
LGTM!

There are some changes in the test files related to formatting that are not desirable, but we can address that in the follow up PR. 

Thank you very much @Soremwar; this PR brings us closer to the goal of Node compatibility. Subtle changes in the bahavior of streams caused some strange bugs that were hard to chase, this PR should alleviate that.

**Pull request merged**